chalk-ir/src/lib.rs

```diff
@@ -411,13 +413,53 @@ impl TraitRef {
     }
 }
 
+/// An inline bound, e.g. `: Foo<K>` in `impl<K, T: Foo<K>> SomeType<T>`.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum InlineBound {
+    TraitBound(TraitBound),
+    ProjectionEqBound(ProjectionEqBound),
+}
+
+pub type QuantifiedInlineBound = Binders<InlineBound>;
+
 /// Where clauses that can be written by a Rust programmer.
 #[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
 pub enum WhereClause {
     Implemented(TraitRef),
     ProjectionEq(ProjectionEq),
 }
 
+/// Represents a trait bound on e.g. a type or type parameter.
+/// Does not know anything about what it's binding.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TraitBound {
+    pub trait_id: ItemId,
+    pub args_no_self: Vec<Parameter>,
```

Comment:
Random thought: we may want to introduce a `Binder` and use that to represent `Self`. I was sort of hoping that we could make a trait object something like

```
exists<Self> { Clause* }
```

Reply:
I see. That could work, what advantages would you see in using this representation?

Reply:
I was thinking that there would be less duplication -- i.e., we wouldn't need a `TraitBound` that is distinct from other clauses.